### PR TITLE
Implement orb--virtual-fields-replacements

### LINKED ISSUE #38: `orb-preformat-keywords`
See discussion in #37 .

### PR DESCRIPTION
**Pull request by zaeph**
Fixes #38.

### CODE REVIEW

**Comment by zaeph** [comment]
Here's an example of the processing `orb--replace-special-keywords` handles:
```el
(orb--replace-special-keywords
 '("citekey" ("my-citekey" . "citekey")))
;; Result:
;; (("citekey" . "=key=")
;;  ("my-citekey" . "=key="))
```

File: org-roam-bibtex.el [hunk 1]
```diff
@@ -120,7 +120,10 @@ See `orb-edit-notes' for details."
           (const :tag "No" nil))
   :group 'org-roam-bibtex)
 
-(defcustom orb-preformat-keywords '(("citekey" . "=key="))
+(defcustom orb-preformat-keywords
+  '("citekey" "type" "pdf?" "note?"
+    "author" "editor"
+    "author-abbrev" "editor-abbrev" "author-or-editor-abbrev")
```
Comment (zaeph) [inline org-roam-bibtex.el:126]: Just to be clear, `author-or-editor-abbrev` doesn't exist yet, but we're working on this in tmalsburg/helm-bibtex/pull/321.

Comment (myshevchuk) [reply org-roam-bibtex.el:126]: Ok.

File: org-roam-bibtex.el [hunk 2]
```diff
@@ -275,14 +278,40 @@ to enter"
             (when (s-equals? p-name (car orb-persp-project))
               (persp-switch p-name))))))))
 
+(defvar orb--virtual-fields-replacements
```
[the remark this answers is not in the document]
Comment (zaeph) [reply org-roam-bibtex.el:281]: Yeah, that was a dumb name, sorry.

File: org-roam-bibtex.el [hunk 3]
```diff
@@ -275,14 +278,40 @@ to enter"
...
+  '(("=type=" . "type")
+    ("=key=" . "citekey")
+    ("=has-pdf=" . "pdf?")
+    ("=has-note=" . "note?"))
+  "Alist of '='-embraced virtual-fields with their replacements.")
+
+(defun orb--replace-special-keywords (kwds)
```
[the remark this answers is not in the document]
Comment (zaeph) [reply org-roam-bibtex.el:288]: Thanks.  Don't know why I've named things so inconsistently.

File: org-roam-bibtex.el [hunk 4]
```diff
@@ -275,14 +278,40 @@ to enter"
...
+  (let* ((replacements orb--virtual-fields-replacements)
+         (fields (mapcar #'cdr replacements)))
+    (mapcar (lambda (kwd)
+              (pcase kwd
+                ((and (pred stringp)
+                      (pred (lambda (x) (member x fields))))
+                 (cons kwd (car (rassoc kwd replacements))))
+                (`(,kwd1 . ,kwd2)
+                 (when (member kwd2 fields)
+                   (cons kwd1 (car (rassoc kwd2 replacements)))))
```
[the remark this answers is not in the document]
Comment (zaeph) [reply org-roam-bibtex.el:301]: I've elected to include the check in the `pcase`, which also allowed me to remove the `when` as I'd done before.  Thanks for spotting this!

File: org-roam-bibtex.el [hunk 5]
```diff
@@ -275,14 +278,40 @@ to enter"
...
+                (_ kwd)))
```
[the remark this answers is not in the document]
Comment (zaeph) [reply org-roam-bibtex.el:302]: I'm working on it.

[the remark this answers is not in the document]
Comment (zaeph) [reply org-roam-bibtex.el]: Implemented in b287d28481c7760f49a584f63f648eb7aa453b0a.

**Comment by zaeph** [comment]
> Just a suggestion, maybe`(_ kwd)` is a good place to throw some errors about a malformed `orb-preformat-keywords`?

b287d28481c7760f49a584f63f648eb7aa453b0a does it.  It prints the following message:
> Wrong type argument: (stringp consp), foo.  Please adjust ‘orb-preformat-keywords’

**Comment by myshevchuk** [comment]
Looks good!

**Comment by zaeph** [comment]
Merging, then!

**Pull request merged**